### PR TITLE
Fix git_submodule_sync with relative url

### PR DESCRIPTION
**Pull request by kdj0c**
git_submodule_sync should resolve submodule before writing to .git/config
to have the same behavior as git_submodule_init, which does the right thing.

This should solve issue #5301

### CODE REVIEW

File: src/submodule.c [hunk 1]
```diff
@@ -1457,12 +1458,13 @@ int git_submodule_sync(git_submodule *sm)
...
 
 		if (!error)
-			error = git_config__update_entry(cfg, key.ptr, sm->url, true, false);
+			error = git_config__update_entry(cfg, key.ptr, effective_submodule_url.ptr, true, false);
 
 		git_repository_free(smrepo);
 	}
 
 	git_buf_dispose(&key);
+    git_buf_dispose(&effective_submodule_url);
```
[the remark this answers is not in the document]
Comment (kdj0c) [reply src/submodule.c]: oh sorry, will fix this asap ;)

**Comment by kdj0c** [comment]
I've modified a test in test_submodule_modify__sync() to check for this behaviour.
The test is failed without the "fix" commit.
I forced submodule3 to be relative, but maybe I should add a new submodule4 in this test ?

**Comment by gdetrez** [comment]
@kdj0c I can confirm that this fixes #5301 when I patch the git2 rust crate with this branch :tada: 

It does break some edge case however: if the superproject has no commits (i.e. it's a freshly initialized repository), running `git_submodule_sync` on a submodule now returns an error: `"reference \'refs/heads/master\' not found"`.
But I'm not sure how important this edge case is.  I only found out because this is what happens with the reproduction steps in #5301.  (This error doesn't happen with the `git submodule sync` command)

**Comment by pks-t** [comment]
@kdj0c: would you feel comfortable with fixing the edge case pointed out by @gdetrez? Let me know if not, then I'll dig into this and amend your PR

**Comment by kdj0c** [comment]
I looked into your edge case, I think the problem is that get_url_base() calls git_repository_head() which returns GIT_EUNBORNBRANCH error.
but as in your test sequence, you are doing "git remote add origin https://github.com/libgit2/libgit2.git", get_url_base() should have found this url and returned it.
It's a bit too complex for me at this point, and I don't want to break other use-case.
@pks-t you can dig deeper, if you think this edge case is important.

**Comment by kdj0c** [comment]
@gdetrez can you try again with my latest patchset ? it "may" solve the edge case.

**Comment by pks-t** [comment]
While your added test exercised the code path, it didn't verify that results are what we'd expect. I've thus replaced it with a new test case that explicitly checks resulting paths and pushed that on your branch. I've naturally retained authorship of your commits. I hope you don't mind that fixup :)

The fix for unborn branches looks good to me, too, so this PR should be ready to go. I'll give it a few days to settle and wait for feedback. If nobody complains, I'll merge it towards the end of this week. Please feel free to reach out to us in case I forget.

**Comment by kdj0c** [comment]
Thanks a lot, I'm looking forward to have it merged ;)

**Comment by pks-t** [comment]
On Mon, Jan 06, 2020 at 12:54:42PM -0800, kdj0c wrote:
> Thanks a lot, I'm looking forward to have it merged ;)

Yeah, I know, that's a fair point. @ethomson, @tiennou: any final
thoughts on this?

**Comment by gdetrez** [comment]
@kdj0c Sorry for the delay.  I tested the latest version (11e8ee1) and it did fix the issue with fresh repository :+1:

**Comment by ethomson** [comment]
LGTM.  Thanks @kdj0c, @gdetrez and @pks-t!
